File: backend/app/routers/problems.py
"""
문제(Problem) API 라우터

Phase 21+ A-2: Problem REST API

문제 CRUD 및 검색 API 엔드포인트
"""

from fastapi import APIRouter, Depends, HTTPException, Query, Body
from typing import List, Optional

from ..models.problem import (
    Problem,
    ProblemCreate,
    ProblemUpdate,
    ProblemFilter,
    ProblemListResponse,
    ProblemStats,
)
from ..services.problems import ProblemService, get_problem_service


router = APIRouter(prefix="/api/problems", tags=["problems"])


# ========== 목록/통계 엔드포인트 (정적 경로 먼저!) ==========

@router.get("", response_model=ProblemListResponse)
async def list_problems(
    # 페이지네이션
    page: int = Query(1, ge=1, description="페이지 번호"),
    pageSize: int = Query(20, ge=1, le=100, description="페이지 크기"),
    # 정렬
    sortBy: str = Query("createdAt", description="정렬 기준"),
    sortDesc: bool = Query(True, description="내림차순 정렬"),
    # 분류 필터
    gradeIds: Optional[str] = Query(None, description="학년 ID 목록 (콤마 구분)"),
    majorUnitIds: Optional[str] = Query(None, description="대단원 ID 목록 (콤마 구분)"),
    middleUnitIds: Optional[str] = Query(None, description="중단원 ID 목록 (콤마 구분)"),
    # 속성 필터
    questionTypes: Optional[str] = Query(None, description="문제 유형 목록 (콤마 구분)"),
    difficultyMin: Optional[int] = Query(None, ge=1, le=10, description="최소 난이도"),
    difficultyMax: Optional[int] = Query(None, ge=1, le=10, description="최대 난이도"),
    # 출처 필터
    sourceTypes: Optional[str] = Query(None, description="출처 유형 목록 (콤마 구분)"),
    years: Optional[str] = Query(None, description="출제 연도 목록 (콤마 구분)"),
    # 기타 필터
    tags: Optional[str] = Query(None, description="태그 목록 (콤마 구분)"),
    hasAnswer: Optional[bool] = Query(None, description="정답 있는 문제만"),
    hasSolution: Optional[bool] = Query(None, description="해설 있는 문제만"),
    isFavorite: Optional[bool] = Query(None, description="즐겨찾기만"),
    # 검색
    searchQuery: Optional[str] = Query(None, description="검색어"),
    # 서비스
    service: ProblemService = Depends(get_problem_service),
):
    """
    문제 목록 조회

    필터링, 정렬, 페이지네이션을 지원합니다.
    """
    # 콤마 구분 문자열을 리스트로 변환
    def parse_int_list(s: Optional[str]) -> Optional[List[int]]:
        if not s:
            return None
        return [int(x.strip()) for x in s.split(",") if x.strip().isdigit()]

    def parse_str_list(s: Optional[str]) -> Optional[List[str]]:
        if not s:
            return None
        return [x.strip() for x in s.split(",") if x.strip()]

    # 필터 구성
    filter_obj = ProblemFilter(
        gradeIds=parse_int_list(gradeIds),
        majorUnitIds=parse_int_list(majorUnitIds),
        middleUnitIds=parse_int_list(middleUnitIds),
        questionTypes=parse_str_list(questionTypes),
        difficultyMin=difficultyMin,
        difficultyMax=difficultyMax,
        sourceTypes=parse_str_list(sourceTypes),
        years=parse_int_list(years),
        tags=parse_str_list(tags),
        hasAnswer=hasAnswer,
        hasSolution=hasSolution,
        isFavorite=isFavorite,
        searchQuery=searchQuery,
    )

    return service.list(
        filter=filter_obj,
        page=page,
        page_size=pageSize,
        sort_by=sortBy,
        sort_desc=sortDesc,
    )


@router.get("/stats/summary", response_model=ProblemStats)
async def get_stats(
    service: ProblemService = Depends(get_problem_service),
):
    """
    문제 통계 조회

    전체 문제 통계를 반환합니다.
    """
    return service.get_stats()


@router.get("/tags/all", response_model=List[str])
async def get_all_tags(
    service: ProblemService = Depends(get_problem_service),
):
    """
    전체 태그 목록 조회

    모든 문제에서 사용된 태그 목록을 반환합니다.
    """
    return service.get_all_tags()


@router.get("/by-source/{document_id}", response_model=List[Problem])
async def get_by_source(
    document_id: str,
    groupId: Optional[str] = Query(None, description="그룹 ID"),
    service: ProblemService = Depends(get_problem_service),
):
    """
    출처 기준 문제 조회

    라벨링 시스템에서 생성된 문제를 조회할 때 사용합니다.
    """
    return service.get_by_source(document_id, groupId)


@router.post("", response_model=Problem)
async def create_problem(
    data: ProblemCreate,
    service: ProblemService = Depends(get_problem_service),
):
    """
    문제 생성

    새로운 문제를 생성합니다.
    """
    return service.create(data)


@router.post("/search", response_model=ProblemListResponse)
async def search_problems(
    filter: ProblemFilter = Body(...),
    page: int = Query(1, ge=1),
    pageSize: int = Query(20, ge=1, le=100),
    sortBy: str = Query("createdAt"),
    sortDesc: bool = Query(True),
    service: ProblemService = Depends(get_problem_service),
):
    """
    문제 검색 (POST)

    복잡한 필터 조건으로 검색할 때 사용합니다.
    """
    return service.list(
        filter=filter,
        page=page,
        page_size=pageSize,
        sort_by=sortBy,
        sort_desc=sortDesc,
    )


@router.post("/bulk", response_model=List[Problem])
async def bulk_create(
    items: List[ProblemCreate],
    service: ProblemService = Depends(get_problem_service),
):
    """
    문제 일괄 생성

    여러 문제를 한 번에 생성합니다.
    """
    return service.bulk_create(items)


@router.post("/bulk-fetch", response_model=List[Problem])
async def bulk_fetch(
    ids: List[str] = Body(..., embed=True),
    service: ProblemService = Depends(get_problem_service),
):
    """
    문제 일괄 조회

    Phase F-1: 여러 문제를 한 번에 조회합니다.
    시험지 미리보기에서 사용됩니다.

    Request Body:
    {
        "ids": ["prob1", "prob2", ...]
    }
    """
    return service.bulk_get(ids)


@router.post("/import-from-labeling", response_model=List[Problem])
async def import_from_labeling(
    data: dict = Body(...),
    service: ProblemService = Depends(get_problem_service),
):
    """
    라벨링 시스템에서 문제 가져오기

    Phase 21+ C-1: 라벨링 → 문제은행 연동

    내보낸 문제 이미지를 문제은행에 등록합니다.

    Request Body:
    {
        "documentId": str,           # 문서 ID
        "documentName": str,         # 문서명 (출처명으로 사용)
        "problems": [                # 가져올 문제 목록
            {
                "group_id": str,
                "page_index": int,
                "image_path": str,
                "column": str
            }
        ],
        "defaultSource": {           # 기본 출처 정보
            "type": "book" | "exam" | "custom",
            "name": str,             # 출처명 (없으면 documentName 사용)
            "year": int?,
            "month": int?,
            "organization": str?
        },
        "classification": {          # 기본 분류 (선택)
            "gradeId": int,
            "majorUnitId": int?,
            ...
        },
        "difficulty": int,           # 기본 난이도 (1-10)
        "questionType": str          # 기본 문제 유형
    }
    """
    from app.config import config

    document_id = data.get("documentId")
    document_name = data.get("documentName", document_id)
    problems_data = data.get("problems", [])
    default_source = data.get("defaultSource", {})
    classification = data.get("classification")
    difficulty = data.get("difficulty", 5)
    question_type = data.get("questionType", "short_answer")

    if not document_id or not problems_data:
        raise HTTPException(
            status_code=400,
            detail="documentId와 problems는 필수입니다"
        )

    # ProblemCreate 목록 생성
    items = []
    for p in problems_data:
        # 이미지 URL 생성 (API 경로로 변환)
        image_path = p.get("image_path", "")
        image_url = f"/api/documents/{document_id}/problems/image?image_path={image_path}"

        # 출처 정보 구성
        source = {
            "type": default_source.get("type", "book"),
            "name": default_source.get("name") or document_name,
            "page": p.get("page_index", 0) + 1,  # 0-indexed → 1-indexed
            "problemNumber": int(p.get("group_id", "0").replace("G", "").replace("g", "")) if p.get("group_id") else None,
            "year": default_source.get("year"),
            "month": default_source.get("month"),
            "organization": default_source.get("organization"),
            "documentId": document_id,  # 원본 문서 ID 저장
            "groupId": p.get("group_id"),  # 원본 그룹 ID 저장
        }

        item = ProblemCreate(
            classification=classification,
            questionType=question_type,
            difficulty=difficulty,
            source=source,
            content={
                "imageUrl": image_url,
                "thumbnailUrl": image_url,
            },
            tags=[],
        )
        items.append(item)

    # 일괄 생성
    return service.bulk_create(items)


@router.post("/import-from-hangul", response_model=List[Problem])
async def import_from_hangul(
    data: dict = Body(...),
    service: ProblemService = Depends(get_problem_service),
):
    """
    한글 파일에서 문제 가져오기

    Phase 21+ C-2: 한글 파일 → 문제은행 연동

    파싱된 한글 문제를 문제은행에 등록합니다.

    Request Body:
    {
        "problems": [                # 파싱된 문제 목록
            {
                "id": str,
                "number": str,
                "content_text": str,
                "content_latex": str,
                "answer": str?,
                "answer_latex": str?,
                "explanation": str?
            }
        ],
        "defaultSource": {           # 기본 출처 정보
            "type": "book" | "exam" | "custom",
            "name": str
        },
        "classification": {          # 기본 분류 (선택)
            "gradeId": int,
            "majorUnitId": int?,
            ...
        },
        "difficulty": int,           # 기본 난이도 (1-10)
        "questionType": str          # 기본 문제 유형
    }
    """
    problems_data = data.get("problems", [])
    default_source = data.get("defaultSource", {})
    classification = data.get("classification")
    difficulty = data.get("difficulty", 5)
    question_type = data.get("questionType", "short_answer")

    if not problems_data:
        raise HTTPException(
            status_code=400,
            detail="problems는 필수입니다"
        )

    # ProblemCreate 목록 생성
    items = []
    for p in problems_data:
        # 출처 정보 구성
        source = {
            "type": default_source.get("type", "book"),
            "name": default_source.get("name", "한글 파일"),
            "problemNumber": int(p.get("number", "0")) if p.get("number", "").isdigit() else None,
        }

        # 정답 및 해설
        answer_text = p.get("answer") or p.get("answer_latex")
        solution_text = p.get("explanation")

        item = ProblemCreate(
            classification=classification,
            questionType=question_type,
            difficulty=difficulty,
            source=source,
            content={
                "text": p.get("content_latex") or p.get("content_text"),
            },
            answer=answer_text,
            solution=solution_text,
            tags=[],
        )
        items.append(item)

    # 일괄 생성
    return service.bulk_create(items)


# ========== 동적 경로 (/{problem_id})는 마지막에! ==========

@router.get("/{problem_id}", response_model=Problem)
async def get_problem(
    problem_id: str,
    service: ProblemService = Depends(get_problem_service),
):
    """
    문제 조회

    ID로 문제를 조회합니다.
    """
    problem = service.get(problem_id)
    if not problem:
        raise HTTPException(status_code=404, detail="문제를 찾을 수 없습니다")
    return problem


@router.patch("/{problem_id}", response_model=Problem)
async def update_problem(
    problem_id: str,
    data: ProblemUpdate,
    service: ProblemService = Depends(get_problem_service),
):
    """
    문제 수정

    문제 정보를 수정합니다.
    """
    problem = service.update(problem_id, data)
    if not problem:
        raise HTTPException(status_code=404, detail="문제를 찾을 수 없습니다")
    return problem


@router.delete("/{problem_id}")
async def delete_problem(
    problem_id: str,
    service: ProblemService = Depends(get_problem_service),
):
    """
    문제 삭제

    문제를 삭제합니다.
    """
    success = service.delete(problem_id)
    if not success:
        raise HTTPException(status_code=404, detail="문제를 찾을 수 없습니다")
    return {"success": True, "message": "문제가 삭제되었습니다"}


@router.post("/{problem_id}/use", response_model=Problem)
async def increment_usage(
    problem_id: str,
    service: ProblemService = Depends(get_problem_service),
):
    """
    문제 사용 횟수 증가

    시험지에 문제가 추가될 때 호출합니다.
    """
    problem = service.increment_usage(problem_id)
    if not problem:
        raise HTTPException(status_code=404, detail="문제를 찾을 수 없습니다")
    return problem


@router.post("/{problem_id}/favorite", response_model=Problem)
async def toggle_favorite(
    problem_id: str,
    service: ProblemService = Depends(get_problem_service),
):
    """
    즐겨찾기 토글

    문제의 즐겨찾기 상태를 토글합니다.
    """
    problem = service.toggle_favorite(problem_id)
    if not problem:
        raise HTTPException(status_code=404, detail="문제를 찾을 수 없습니다")
    return problem

File: frontend/src/api/problems.ts
/**
 * 문제(Problem) API 클라이언트
 *
 * Phase 21+ A-2: Problem API 훅
 */

import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query';
import type {
  Problem,
  ProblemCreate,
  ProblemUpdate,
  ProblemFilter,
  ProblemListResponse,
  ProblemStats,
  ProblemSortOptions,
} from '../types/problem';

const API_BASE = 'http://localhost:8000/api/problems';

// ========== API 함수 ==========

async function fetchProblems(params: {
  page?: number;
  pageSize?: number;
  sortBy?: string;
  sortDesc?: boolean;
  filter?: Partial<ProblemFilter>;
}): Promise<ProblemListResponse> {
  const searchParams = new URLSearchParams();

  if (params.page) searchParams.set('page', String(params.page));
  if (params.pageSize) searchParams.set('pageSize', String(params.pageSize));
  if (params.sortBy) searchParams.set('sortBy', params.sortBy);
  if (params.sortDesc !== undefined) searchParams.set('sortDesc', String(params.sortDesc));

  // 필터 파라미터
  const filter = params.filter || {};
  if (filter.gradeIds?.length) searchParams.set('gradeIds', filter.gradeIds.join(','));
  if (filter.majorUnitIds?.length) searchParams.set('majorUnitIds', filter.majorUnitIds.join(','));
  if (filter.middleUnitIds?.length) searchParams.set('middleUnitIds', filter.middleUnitIds.join(','));
  if (filter.questionTypes?.length) searchParams.set('questionTypes', filter.questionTypes.join(','));
  if (filter.difficultyMin !== undefined) searchParams.set('difficultyMin', String(filter.difficultyMin));
  if (filter.difficultyMax !== undefined) searchParams.set('difficultyMax', String(filter.difficultyMax));
  if (filter.sourceTypes?.length) searchParams.set('sourceTypes', filter.sourceTypes.join(','));
  if (filter.years?.length) searchParams.set('years', filter.years.join(','));
  if (filter.tags?.length) searchParams.set('tags', filter.tags.join(','));
  if (filter.hasAnswer !== undefined) searchParams.set('hasAnswer', String(filter.hasAnswer));
  if (filter.hasSolution !== undefined) searchParams.set('hasSolution', String(filter.hasSolution));
  if (filter.isFavorite !== undefined) searchParams.set('isFavorite', String(filter.isFavorite));
  if (filter.searchQuery) searchParams.set('searchQuery', filter.searchQuery);

  const response = await fetch(`${API_BASE}?${searchParams}`);
  if (!response.ok) throw new Error('문제 목록 조회 실패');
  return response.json();
}

async function fetchProblem(id: string): Promise<Problem> {
  const response = await fetch(`${API_BASE}/${id}`);
  if (!response.ok) throw new Error('문제 조회 실패');
  return response.json();
}

async function createProblem(data: ProblemCreate): Promise<Problem> {
  const response = await fetch(API_BASE, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(data),
  });
  if (!response.ok) throw new Error('문제 생성 실패');
  return response.json();
}

async function updateProblem(id: string, data: ProblemUpdate): Promise<Problem> {
  const response = await fetch(`${API_BASE}/${id}`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(data),
  });
  if (!response.ok) throw new Error('문제 수정 실패');
  return response.json();
}

async function deleteProblem(id: string): Promise<void> {
  const response = await fetch(`${API_BASE}/${id}`, {
    method: 'DELETE',
  });
  if (!response.ok) throw new Error('문제 삭제 실패');
}

async function fetchProblemStats(): Promise<ProblemStats> {
  const response = await fetch(`${API_BASE}/stats/summary`);
  if (!response.ok) throw new Error('통계 조회 실패');
  return response.json();
}

async function fetchAllTags(): Promise<string[]> {
  const response = await fetch(`${API_BASE}/tags/all`);
  if (!response.ok) throw new Error('태그 조회 실패');
  return response.json();
}

async function toggleFavorite(id: string): Promise<Problem> {
  const response = await fetch(`${API_BASE}/${id}/favorite`, {
    method: 'POST',
  });
  if (!response.ok) throw new Error('즐겨찾기 토글 실패');
  return response.json();
}

async function incrementUsage(id: string): Promise<Problem> {
  const response = await fetch(`${API_BASE}/${id}/use`, {
    method: 'POST',
  });
  if (!response.ok) throw new Error('사용 횟수 증가 실패');
  return response.json();
}

async function bulkCreate(items: ProblemCreate[]): Promise<Problem[]> {
  const response = await fetch(`${API_BASE}/bulk`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(items),
  });
  if (!response.ok) throw new Error('일괄 생성 실패');
  return response.json();
}

async function fetchBySource(documentId: string, groupId?: string): Promise<Problem[]> {
  const url = groupId
    ? `${API_BASE}/by-source/${documentId}?groupId=${groupId}`
    : `${API_BASE}/by-source/${documentId}`;
  const response = await fetch(url);
  if (!response.ok) throw new Error('출처별 조회 실패');
  return response.json();
}

/**
 * Phase F-1: 문제 일괄 조회
 */
async function bulkFetchProblems(ids: string[]): Promise<Problem[]> {
  if (ids.length === 0) return [];

  const response = await fetch(`${API_BASE}/bulk-fetch`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ ids }),
  });
  if (!response.ok) throw new Error('일괄 조회 실패');
  return response.json();
}

async function searchProblems(
  filter: ProblemFilter,
  page = 1,
  pageSize = 20,
  sortBy = 'createdAt',
  sortDesc = true,
): Promise<ProblemListResponse> {
  const params = new URLSearchParams({
    page: String(page),
    pageSize: String(pageSize),
    sortBy,
    sortDesc: String(sortDesc),
  });

  const response = await fetch(`${API_BASE}/search?${params}`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(filter),
  });
  if (!response.ok) throw new Error('검색 실패');
  return response.json();
}

/**
 * Phase 21+ C-1: 라벨링에서 문제은행으로 가져오기
 */
export interface ImportFromLabelingParams {
  documentId: string;
  documentName: string;
  problems: Array<{
    group_id: string;
    page_index: number;
    image_path: string;
    column?: string;
  }>;
  defaultSource?: {
    type?: 'book' | 'exam' | 'custom';
    name?: string;
    year?: number;
    month?: number;
    organization?: string;
  };
  classification?: {
    gradeId?: number;
    majorUnitId?: number;
    middleUnitId?: number;
    minorUnitId?: number;
    typeId?: number;
  };
  difficulty?: number;
  questionType?: 'multiple_choice' | 'short_answer' | 'essay';
}

async function importFromLabeling(params: ImportFromLabelingParams): Promise<Problem[]> {
  const response = await fetch(`${API_BASE}/import-from-labeling`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(params),
  });
  if (!response.ok) throw new Error('라벨링 가져오기 실패');
  return response.json();
}

// ========== Query Keys ==========

export const problemKeys = {
  all: ['problems'] as const,
  lists: () => [...problemKeys.all, 'list'] as const,
  list: (params: object) => [...problemKeys.lists(), params] as const,
  details: () => [...problemKeys.all, 'detail'] as const,
  detail: (id: string) => [...problemKeys.details(), id] as const,
  stats: () => [...problemKeys.all, 'stats'] as const,
  tags: () => [...problemKeys.all, 'tags'] as const,
  bySource: (documentId: string, groupId?: string) =>
    [...problemKeys.all, 'bySource', documentId, groupId] as const,
  bulk: (ids: string[]) => [...problemKeys.all, 'bulk', ids.sort().join(',')] as const,
};

// ========== Query Hooks ==========

/**
 * 문제 목록 조회
 */
export function useProblems(params: {
  page?: number;
  pageSize?: number;
  sortBy?: string;
  sortDesc?: boolean;
  filter?: Partial<ProblemFilter>;
  enabled?: boolean;
} = {}) {
  const { enabled = true, ...queryParams } = params;

  return useQuery({
    queryKey: problemKeys.list(queryParams),
    queryFn: () => fetchProblems(queryParams),
    enabled,
  });
}

/**
 * 단일 문제 조회
 */
export function useProblem(id: string | undefined) {
  return useQuery({
    queryKey: problemKeys.detail(id || ''),
    queryFn: () => fetchProblem(id!),
    enabled: !!id,
  });
}

/**
 * 문제 통계 조회
 */
export function useProblemStats() {
  return useQuery({
    queryKey: problemKeys.stats(),
    queryFn: fetchProblemStats,
  });
}

/**
 * 전체 태그 목록 조회
 */
export function useAllTags() {
  return useQuery({
    queryKey: problemKeys.tags(),
    queryFn: fetchAllTags,
  });
}

/**
 * 출처별 문제 조회
 */
export function useProblemsBySource(documentId: string, groupId?: string) {
  return useQuery({
    queryKey: problemKeys.bySource(documentId, groupId),
    queryFn: () => fetchBySource(documentId, groupId),
    enabled: !!documentId,
  });
}

/**
 * Phase F-1: 문제 일괄 조회
 *
 * 시험지 미리보기에서 여러 문제를 한 번에 조회할 때 사용
 */
export function useBulkProblems(ids: string[], enabled = true) {
  return useQuery({
    queryKey: problemKeys.bulk(ids),
    queryFn: () => bulkFetchProblems(ids),
    enabled: enabled && ids.length > 0,
  });
}

// ========== Mutation Hooks ==========

/**
 * 문제 생성
 */
export function useCreateProblem() {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: createProblem,
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: problemKeys.lists() });
      queryClient.invalidateQueries({ queryKey: problemKeys.stats() });
      queryClient.invalidateQueries({ queryKey: problemKeys.tags() });
    },
  });
}

/**
 * 문제 수정
 */
export function useUpdateProblem() {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: ({ id, data }: { id: string; data: ProblemUpdate }) =>
      updateProblem(id, data),
    onSuccess: (_, { id }) => {
      queryClient.invalidateQueries({ queryKey: problemKeys.detail(id) });
      queryClient.invalidateQueries({ queryKey: problemKeys.lists() });
      queryClient.invalidateQueries({ queryKey: problemKeys.stats() });
    },
  });
}

/**
 * 문제 삭제
 */
export function useDeleteProblem() {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: deleteProblem,
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: problemKeys.lists() });
      queryClient.invalidateQueries({ queryKey: problemKeys.stats() });
    },
  });
}

/**
 * 즐겨찾기 토글
 */
export function useToggleFavorite() {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: toggleFavorite,
    onSuccess: (problem) => {
      queryClient.setQueryData(problemKeys.detail(problem.id), problem);
      queryClient.invalidateQueries({ queryKey: problemKeys.lists() });
      queryClient.invalidateQueries({ queryKey: problemKeys.stats() });
    },
  });
}

/**
 * 사용 횟수 증가
 */
export function useIncrementUsage() {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: incrementUsage,
    onSuccess: (problem) => {
      queryClient.setQueryData(problemKeys.detail(problem.id), problem);
    },
  });
}

/**
 * 문제 일괄 생성
 */
export function useBulkCreateProblems() {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: bulkCreate,
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: problemKeys.lists() });
      queryClient.invalidateQueries({ queryKey: problemKeys.stats() });
      queryClient.invalidateQueries({ queryKey: problemKeys.tags() });
    },
  });
}

/**
 * 문제 검색 (POST)
 */
export function useSearchProblems() {
  return useMutation({
    mutationFn: ({
      filter,
      page,
      pageSize,
      sortBy,
      sortDesc,
    }: {
      filter: ProblemFilter;
      page?: number;
      pageSize?: number;
      sortBy?: string;
      sortDesc?: boolean;
    }) => searchProblems(filter, page, pageSize, sortBy, sortDesc),
  });
}

/**
 * Phase 21+ C-1: 라벨링에서 문제은행으로 가져오기
 */
export function useImportFromLabeling() {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: importFromLabeling,
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: problemKeys.lists() });
      queryClient.invalidateQueries({ queryKey: problemKeys.stats() });
      queryClient.invalidateQueries({ queryKey: problemKeys.tags() });
    },
  });
}

/**
 * Phase 21+ C-2: 한글 파일에서 문제은행으로 가져오기
 */
export interface ImportFromHangulParams {
  problems: Array<{
    id: string;
    number: string;
    content_text: string;
    content_latex: string;
    answer?: string | null;
    answer_latex?: string | null;
    explanation?: string | null;
  }>;
  defaultSource?: {
    type?: 'book' | 'exam' | 'custom';
    name?: string;
  };
  classification?: {
    gradeId?: number;
    majorUnitId?: number;
    middleUnitId?: number;
    minorUnitId?: number;
    typeId?: number;
  };
  difficulty?: number;
  questionType?: 'multiple_choice' | 'short_answer' | 'essay';
}

async function importFromHangul(params: ImportFromHangulParams): Promise<Problem[]> {
  const response = await fetch(`${API_BASE}/import-from-hangul`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(params),
  });
  if (!response.ok) throw new Error('한글 파일 가져오기 실패');
  return response.json();
}

export function useImportFromHangul() {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: importFromHangul,
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: problemKeys.lists() });
      queryClient.invalidateQueries({ queryKey: problemKeys.stats() });
      queryClient.invalidateQueries({ queryKey: problemKeys.tags() });
    },
  });
}
